routes/artifacts.js

```diff
@@ -0,0 +1,597 @@
+    details.message   = input.message;
+    details.reason    = input.reason;
+  }
+
+  try {
+    var artifact = await this.Artifact.create({
+      taskId:           taskId,
+      runId:            runId,
+      name:             name,
+      storageType:      storageType,
+      contentType:      contentType,
+      details:          details,
+      expires:          expires
+    });
+  }
+  catch (err) {
+    // Re-throw error if this isn't because the entity already exists
+    if (!err || err.code !== 'EntityAlreadyExists') {
+      throw err;
+    }
+
+    // Load existing Artifact entity
+    artifact = await this.Artifact.load({
+      taskId:           taskId,
+      runId:            runId,
+      name:             name
+    });
+
+    // Allow recreating of the same artifact, report conflict if it's not the
+    // same artifact (allow for later expiration).
+    // Note, we'll check `details` later
+    if (artifact.storageType !== storageType ||
+        artifact.contentType !== contentType ||
+        artifact.expires.getTime() > expires.getTime()) {
+      return res.status(409).json({
+        message:  "Artifact already exists, with different type or " +
+                  " later expiration"
+      });
+    }
+
+    // Check that details match, unless this has storageType 'reference', we'll
+    // workers to overwrite redirect artifacts
+    if (storageType !== 'reference' &&
+        !_.isEqual(artifact.details, details)) {
+      return res.status(409).json({
+        message:  "Artifact already exists with different contentType or " +
+                  "error message"
+      });
+    }
+
+    // Update expiration and detail, which may have been modified
+    await artifact.modify((artifact) => {
+      artifact.expires = expires;
+      artifact.details = details;
+    });
+  }
+
+  // Publish message about artifact creation
+  await this.publisher.artifactCreated({
+    status:         task.status(),
+    artifact:       artifact.json(),
+    workerGroup:    workerGroup,
+    workerId:       workerId,
+    runId:          runId
+  }, task.routes);
+
+  // Reply with signed S3 URL
+  if (artifact.storageType === 's3') {
+    var expiry = new Date(new Date().getTime() + 30 * 60 * 1000);
+    var bucket = null;
+    if (artifact.details.bucket === this.publicBucket.bucket) {
+      bucket = this.publicBucket;
+    }
+    if (artifact.details.bucket === this.privateBucket.bucket) {
+      bucket = this.privateBucket;
+    }
+    // Create put URL
+    var putUrl = await this.privateBucket.createPutUrl(
+      artifact.details.prefix, {
+      contentType:      artifact.contentType,
+      expires:          30 * 60 + 10 // Add 10 sec for clock drift
+    });
+    return res.reply({
+      storageType:  's3',
+      contentType:  artifact.contentType,
+      expires:      expiry.toJSON(),
+      putUrl:       putUrl
+    });
+  }
+
+  // Reply with SAS for azure
+  if (artifact.storageType === 'azure') {
```

Comment:
Moved to switch but really sure it's any better...

I like the return early pattern, and use else and elseif unless I strictly have to...